### DISCUSSION TITLE
Come convertire la lettura del A0 Wemos D1

### DISCUSSION THREAD

**Discussion by kfrancesco** [indomus/forum#766] [HUB: Home Assistant]
Salve a tutti,
sono abbastanza nuovo con Home Assistant ed avrei bisogno di una mano, vi spiego il problema.
Ho sul tetto di casa un pannello solare termico con bolier ed avendo la necessità di monitorare la temperatura dell'acqua ho realizzato un circuito con Wemos D1 ed un paio di sensori DS18B20 (di elettronica ne mastico abbastanza).
Purtroppo non avendo una presa di corrente facilmente accessibile sul tetto di casa, ho dovuto ricorrere ad una alimentazione a una batteria con ricaricata tramite un piccolo pannello solare (ed anche qui niente problemi). Lo stato della batteria è monitorizzato tramite il pin A0 del Wemos D1 (non sarà il massimo della precisone ma per quello che mi serve va benissimo).
Su HA il dispositivo viene visto come Tasmota e su Lovelave riesco a leggere tutti i valori, il problema è che giustamente l'entrata A0 viene visualizzata con valori da 0 a 1024, come la converto in tensione? Ho letto un po' in giro ma mi sono perso.
Grazie

**Comment by xyzmauro** [comment]
Ciao, ipotizzando una batteria da 5v su A0 lèggerai 1023 quando è carica e 0 quando scarica, da qui fai la proporzione…

int sensorValue = analogRead(A0);
float voltage= sensorValue * (5.0 / 1023.0);

**Comment by kfrancesco** [comment]
CIao,
grazie, ma perdonami sono stato poco chiaro e troppo sintetico, non riesco a capire come inserire la conversione nel file di configuration.yaml (se e' qui che va inserita). Ad esempio se metto una cosa del genere il valore è sconosciuto

  - platform: mqtt
    name: "TensioneBatteriaPannelloSolare"
    state_topic: "sensor.sensore_pannellosolare_analog_a0"
    unit_of_measurement: "V"
    value_template: "{{ value_json.a0 }}"

**Reply by kfrancesco** [reply]
Dimenticavo c'é un modo per forzare un eventuale visualizzazione dell'errore? Ad esempio mi dimentico una lettera "sensor.sensore_pannellosolare_alog_a0" al posto di "sensor.sensore_pannellosolare_analog_a0" non c'e' un sistema per dirmi guarda che quel sensore non esiste?
Grazie

**Reply by xyzmauro** [reply]
Mi sa che stai facendo un pò di confusione, se stai facendo un MQTT sensor lo state_topic sarà il topic che il sensore deve andare a leggere, una cosa tipo "tele/sonoff/sensor" mentre il value_template dipende da come è il payload, scritto così è un json e ha "attributo" a0...ma non credo sia questo il caso
Se hai già il sensor.sensore_pannellosolare_analog_a0 che avrà stato da 0 a 1023 allora devi fare un sensor_template dove fare il calcolo
```
- platform: template
  sensors:
    voltage:      
      value_template: >-
        "{{ (states('sensor.sensore_pannellosolare_analog_a0') |float * (5.0 / 1023.0)) |round(2) }}"
```
Spero di aver capito giusto stavolta, il 1° suggerimento partiva dal presupposto che avessi scritto il firmware del Wemos con Arduino, chissà perchè ;-)

**Comment by kfrancesco** [comment]
![Scherm](https://user-images.githubusercontent.com/59313284/134951592-5f21b2a6-729c-4e08-a01b-12e63149e331.png)

Probabilmente hai ragione sto sbagliando sicuramente qualcosa, allora sul Wemos D1 è stato caricato con il tasmota 9.5 ed interfacciato con HA direttamente via MQTT (mi è parsa la via più semplice). Dallo screen sopra puoi vedere che funziona il tutto perfettamente.
Ora andando ad osservare i componenti ho notato che l'integrazione è passata via Tasmota e non con il Mosquitto broker, anche se non capisco il perché visto nello stesso sistema ho un altro Wemos D1 (stesso firmware) con PZEM che monitorizza i consumi ed è visto da MQTT.

**Reply by xyzmauro** [reply]
Probabilmente avrai impostato diversamente  l’autodiscovery tra i due Wemos, se non ricordo male SetOption19…
comunque poco male,se quel 857 che si vede nello screen è il sensor.sensore_pannellosolare_analog_a0 (controlla come lo hai scritto perché sembra diverso) fai il sensor_template che ho messo sopra e sei a posto 😉

**Comment by kfrancesco** [comment]
Graie mille ora funziona perfettamente.
![Schermata](https://user-images.githubusercontent.com/59313284/135072337-8cc2f368-c775-4f53-8df8-9c2495fe343a.png)

Ti posso gentilmente chiedere solo un chiarimento? Non capisco dove sbagliavo, ad esempio per leggere la temperatura di uno shelly faccio 
`  - platform: mqtt
    name: "Temperatura shelly Avvolgibile Finestra"
    state_topic: "shellies/shellyswitch25-68C63A/temperature"
    qos: 0
    unit_of_measurement: "°C"`

Perchè in questo caso ho dovuto ricorrere a 
`  - platform: template
    sensors:
      voltage: ecc ecc`

Non capisco il motivo, grazie

**Reply by xyzmauro** [reply]
Mah,in questo momento tu hai un sensor_mqtt che ti da il valore analogico di A0 che ti passa il Wemos e hai aggiunto un sensor_template che è derivato da quello dove fai il calcolo per trasformarlo in Volt
Se non ti interessa il valore analogico potresti fare tutto in una volta sola ma x il come dipende da come è scritto il JSON nel payload
Altra strada,non conosco abbastanza Tasmota,ma credo che potresti usare un template già lì e far pubblicare direttamente il valore in Volt, su questo non saprei aiutarti, avrei da studiare al posto tuo 😅
Last but not least, potresti lasciar da parte Tasmota e scriverti il firmware come vuoi tu con ESPHome

**Reply by Scialla** [reply]
Templating Mauro... :)
